ARROW-12325: [C++] [CI] Nightly gandiva build failing due to failure of compiler to move return value

westonpace opened the pull request:
Gandiva build failure caused by failure of GCC 4.8.2 compiler to automatically move return value.

Failed build: https://github.com/ursacomputing/crossbow/runs/2303374510
Godbolt showing I'm not crazy: https://gcc.godbolt.org/z/x138zevhE

nealrichardson commented:
> I'm open to suggestions

Maybe add a comment to this effect so that whenever we remove support for gcc < 4.9, we can remove this? Or could you even use an `#ifdef` or something to only do this on gcc < 4.9?

lidavidm commented:
I agree, an ifdef could be OK, though right now this is only used in tests.

I submitted https://github.com/lidavidm/crossbow/actions/runs/748817979 for now.

cyb70289 commented on cpp/src/arrow/util/vector.h:
```diff
@@ -130,7 +130,7 @@ Result<std::vector<T>> UnwrapOrRaise(std::vector<Result<T>>&& results) {
     out.push_back(it->MoveValueUnsafe());
   }
-  return out;
+  return std::move(out);
```
Looks to me this `std::move` does no harm.
Regarding RVO, as we are returning `Result` wrapped object, not the object itself, to my knowledge, there must be a temporary object to be moved or copied by `Result`.

cyb70289 replied:
cc @pitrou @emkornfield

westonpace replied:
I agree, it's harmless and not actively used, so this one instance is not really something to worry over much about.  Thinking on it more I think I was more surprised we get away with it everywhere else.  We are returning `Result` most of the time and we don't usually move return values.

pitrou replied:
Agreed with @cyb70289 . RVO is only guaranteed if the returned value has the exact return type declared by the function.

cyb70289 replied:
@westonpace , also change line 95? And maybe tweak commit message a bit?
I do see `return std::move(local_obj)` is used extensively when returning `Result` in other source files.

westonpace commented:
I'll side with @cyb70289 and @lidavidm that an ifdef would probably be overkill.  Feel free to merge.

kiszk commented:
ifdef looks more clear to show our intention.

westonpace commented:
@cyb70289 I updated line 95 and fixed up the commit message.  @pitrou is correct that we do not need to worry about RVO if the output type differs so it should be harmless (i.e. no need for an ifdef).

kou commented:
Can we merge this to include this into 4.0.0 RC2?

westonpace commented:
The build failures appear to all be expected.  MacOS is failing due to LLVM (ARROW-12467).  C GLib & Ruby is failing on Windows with segmentation fault but was doing so before this (we should probably create a JIRA).  The JNI failure is ARROW-11633
